File: Tarea02/dawkins.js
/**
 * @constant {string} Alfabeto válido: letras mayúsculas A-Z más el espacio. 
 * Los caracteres de todas cadenas deben pertenecer a este alfabeto.
 */
const ABECEDARIO = "ABCDEFGHIJKLMNOPQRSTUVWXYZ ";

/** @type {number} Número de generaciones completadas desde el inicio. */
let generacion = 0;

/** @type {string} Cadena con mayor puntaje en la generación actual. */
let mejorCadena = "";

/** @type {boolean} Indica si hay una simulación en curso; evita ejecuciones simultáneas. */
let enProceso = false;

/** @type {string} Frase objetivo normalizada (mayúsculas, sin caracteres inválidos). */
let objetivo = "";

/** @type {number} Longitud de la frase objetivo; se cachea para evitar recalcularla. */
let objLen = 0;

/** @type {number|null} ID del setTimeout activo; permite cancelar el ciclo si fuera necesario. */
let timeoutId = null;

const entradaObjetivo = document.getElementById("entradaObjetivo");
const ayudaObjetivo = document.getElementById("ayudaObjetivo");
const opcionAleatoria = document.getElementById("opcionAleatoria");
const opcionManual = document.getElementById("opcionManual");
const entradaCadena = document.getElementById("entradaCadena");
const ayudaManual = document.getElementById("ayudaManual");
const btnIniciar = document.getElementById("btnIniciar");
const txtSalida = document.getElementById("txtSalida");

entradaObjetivo.addEventListener("focus", () => {
    ayudaObjetivo.classList.add("texto-resaltado");
});
entradaObjetivo.addEventListener("blur", () => {
    ayudaObjetivo.classList.remove("texto-resaltado");
});

/**
 * Habilita o deshabilita el campo de cadena manual según la opción
 * de radio activa, y gestiona el resaltado del texto de ayuda asociado.
 */
function actualizarModalidad() {
    if (opcionManual.checked) {
        entradaCadena.disabled = false;
        entradaCadena.focus();
        ayudaManual.classList.add("texto-resaltado");
    } else {
        entradaCadena.disabled = true;
        ayudaManual.classList.remove("texto-resaltado");
    }
}

opcionAleatoria.addEventListener("change", actualizarModalidad);
opcionManual.addEventListener("change", actualizarModalidad);

/**
 * Convierte la entrada a mayúsculas y filtra todo carácter que no sea
 * una letra A–Z ni un espacio. Se aplica tanto a la cadena objetivo
 * como a la cadena de inicio manual.
 *
 * @param   {string} cadena Texto de entrada sin restricciones.
 * @returns {string} Cadena en mayúsculas con solo caracteres válidos.
 */
function limpiarCadena(cadena) {
    return cadena.toUpperCase().split('').filter(c => ABECEDARIO.includes(c)).join('');
}

/**
 * Genera una cadena aleatoria del mismo largo que la frase objetivo.
 *
 * @returns {string} Cadena aleatoria de longitud {@link objLen}.
 */
function generarAleatoria() {
    let resultado = "";
    for (let i = 0; i < objLen; i++) {
        const indiceAleatorio = Math.floor(Math.random() * ABECEDARIO.length);
        resultado += ABECEDARIO[indiceAleatorio];
    }
    return resultado;
}

/**
 * Calcula el puntaje de una cadena como el número de posiciones en que 
 * coincide con la cadena objetivo.
 * @param {string} cadena Cadena candidata a puntuar. 
 * @returns {number} Número de caracteres que coinciden posición a posición con {@link objetivo}
 */
function puntaje(cadena) {
    let puntos = 0;
    for (let i = 0; i < objLen; i++) {
        if (cadena[i] === objetivo[i]) {
            puntos++;
        }
    }
    return puntos;
}

/**
 * Aplica mutación aleatoria a cada carácter de la cadena con una probabilidad
 * del 5% por posición.
 * @param   {string} cadena Cadena padre sobre la que se aplican las mutaciones.
 * @returns {string} Nueva cadena resultante de la mutación.
 */
function mutar(cadena) {
    let resultado = "";
    for (let i = 0; i < cadena.length; i++) {
        if (Math.random() < 0.05) { 
            const indiceAleatorio = Math.floor(Math.random() * ABECEDARIO.length);
            resultado += ABECEDARIO[indiceAleatorio];
        } else {
            resultado += cadena[i];
        }
    }
    return resultado;
}

btnIniciar.addEventListener("click", () => {
    if (enProceso) return;

    txtSalida.value = ""; 

    objetivo = limpiarCadena(entradaObjetivo.value);
    objLen = objetivo.length;

    if (objLen === 0) {
        alert("La frase objetivo no puede estar vacía o tener puros caracteres inválidos.");
        return;
    }

    entradaObjetivo.value = objetivo; 

    if (opcionAleatoria.checked) {
        mejorCadena = generarAleatoria();
    } else {
        let cadenaUsuario = limpiarCadena(entradaCadena.value);
        if (cadenaUsuario.length !== objLen) {
            alert(`La cadena debe tener exactamente ${objLen} caracteres válidos.\nTu cadena tiene ${cadenaUsuario.length} caracteres.`);
            return;
        }
        mejorCadena = cadenaUsuario;
    }

    generacion = 0;
    enProceso = true;
    btnIniciar.disabled = true;
    entradaObjetivo.disabled = true;

    evolucionar();
});

/**
 * Ejecuta una generación del algoritmo y programa la siguiente.
 */
function evolucionar() {
    if (!enProceso) return;

    const puntajeActual = puntaje(mejorCadena);
    
    const genFormateada = generacion.toString().padStart(3, '0');
    
    txtSalida.value += `Gen ${genFormateada} | ${mejorCadena} | Puntos: ${puntajeActual}\n`;
    txtSalida.scrollTop = txtSalida.scrollHeight;

    if (puntajeActual === objLen) {
        enProceso = false;
        btnIniciar.disabled = false;
        entradaObjetivo.disabled = false;
        txtSalida.value += `\n¡Evolución completada con éxito en ${generacion} generaciones!\n`;
        return;
    }

    let poblacion = [mejorCadena]; 
    for (let i = 0; i < 100; i++) {
        poblacion.push(mutar(mejorCadena));
    }

    mejorCadena = poblacion.reduce((mejor, actual) => {
        return puntaje(actual) > puntaje(mejor) ? actual : mejor;
    });

    generacion++;

    timeoutId = setTimeout(evolucionar, 10);
}

File: Tarea02/dawkins.py
import random
import time
import sys

# Alfabeto válido: letras mayúsculas A-Z más el espacio.
ABECEDARIO = "ABCDEFGHIJKLMNOPQRSTUVWXYZ "

class MonoEscritor:
    """
    Clase que encapsula el algoritmo de Dawkins para ejecutarse en terminal.
    """
    def __init__(self):
        self.objetivo = "METHINKS IT IS LIKE A WEASEL"
        self.obj_len = len(self.objetivo)
        self.mejor_cadena = ""
        self.generacion = 0

    def limpiar_cadena(self, cadena: str) -> str:
        """Filtra la cadena para dejar solo letras mayúsculas y espacios."""
        return ''.join(c for c in cadena.upper() if c in ABECEDARIO)

    def configurar_objetivo(self):
        """Permite al usuario cambiar el objetivo o dejar el de por defecto."""
        print("\n--- CONFIGURACIÓN DEL OBJETIVO ---")
        print(f"Objetivo actual por defecto: '{self.objetivo}'")
        nuevo = input("Presiona Enter para usar el actual, o escribe una nueva frase objetivo: ")
        
        if nuevo.strip() != "":
            nuevo_limpio = self.limpiar_cadena(nuevo)
            if len(nuevo_limpio) > 0:
                self.objetivo = nuevo_limpio
                self.obj_len = len(self.objetivo)
                print(f"-> Nuevo objetivo configurado: '{self.objetivo}'")
            else:
                print("-> Entrada inválida. Se mantendrá el objetivo por defecto.")
        else:
            print("-> Se usará el objetivo por defecto.")

    def cadena_random(self) -> str:
        """Genera una cadena aleatoria de longitud igual a la del objetivo."""
        return ''.join(random.choice(ABECEDARIO) for _ in range(self.obj_len))

    def puntaje(self, cadena: str) -> int:
        """Calcula cuántas letras coinciden en su posición exacta con el objetivo."""
        return sum(1 for a, b in zip(cadena, self.objetivo) if a == b)

    def mutar(self, cadena: str) -> str:
        """Aplica un 5% de probabilidad de mutación a cada carácter."""
        resultado = ""
        for char in cadena:
            if random.random() < 0.05:
                resultado += random.choice(ABECEDARIO)
            else:
                resultado += char
        return resultado

    def configurar_inicio(self):
        """Permite al usuario elegir entre cadena aleatoria o manual."""
        print("\n--- CONFIGURACIÓN DE INICIO ---")
        print("1. Generar cadena inicial aleatoria")
        print("2. Ingresar cadena inicial manualmente")
        
        while True:
            opcion = input("Elige una opción (1 o 2): ").strip()
            
            if opcion == '1':
                self.mejor_cadena = self.cadena_random()
                break
            elif opcion == '2':
                print(f"\nTu cadena debe tener EXACTAMENTE {self.obj_len} caracteres válidos (letras y espacios).")
                while True:
                    cadena_usr = input(f"Ingresa tu cadena de {self.obj_len} caracteres: ")
                    cadena_limpia = self.limpiar_cadena(cadena_usr)
                    
                    if len(cadena_limpia) == self.obj_len:
                        self.mejor_cadena = cadena_limpia
                        break
                    else:
                        print(f"[Error] Tu cadena tiene {len(cadena_limpia)} caracteres válidos. Intenta de nuevo.")
                break
            else:
                print("[Error] Opción no válida. Por favor, escribe 1 o 2.")

    def evolucionar(self):
        """Ciclo principal de evolución en terminal."""
        print("\n================ INICIANDO EVOLUCIÓN ================\n")
        self.generacion = 0
        
        # Pequeño retraso para que el usuario alcance a leer
        time.sleep(1)

        while True:
            puntaje_actual = self.puntaje(self.mejor_cadena)
            
            # Imprimir estado de la generación (usamos \r para sobrescribir la línea si queremos, 
            # pero imprimir línea por línea da ese efecto "Matrix" de evolución).
            # Formateamos con ceros a la izquierda para que se vea alineado.
            print(f"Gen {self.generacion:04d} | {self.mejor_cadena} | Puntos: {puntaje_actual}")

            # Condición de victoria
            if puntaje_actual == self.obj_len:
                print("\n=====================================================")
                print(f"¡EVOLUCIÓN COMPLETADA CON ÉXITO EN {self.generacion} GENERACIONES!")
                print("=====================================================\n")
                break

            # Crear población de 100 hijos + el padre
            poblacion = [self.mutar(self.mejor_cadena) for _ in range(100)]
            poblacion.append(self.mejor_cadena) # Elitismo

            # Seleccionar al mejor
            self.mejor_cadena = max(poblacion, key=self.puntaje)
            self.generacion += 1

            # Pausa minúscula opcional para que la terminal no escupa todo de golpe (puedes comentarla para máxima velocidad)
            time.sleep(0.01) 

    def iniciar(self):
        """Método principal que orquesta la ejecución."""
        print("=========================================")
        print("   PROGRAMA DE RICHARD DAWKINS (TERMINAL)")
        print("=========================================")
        
        self.configurar_objetivo()
        self.configurar_inicio()
        self.evolucionar()

if __name__ == "__main__":
    juego = MonoEscritor()
    try:
        juego.iniciar()
    except KeyboardInterrupt:
        print